File: desafio4/go/hufftree.go
package main

import("fmt")

const MAX_SYMBOLS = 256
type Tree interface {
	Freq() uint
	WriteTo(writer *BitOutputStream)
	DumpCodes(codes []string, prefix string)
	ReadChar(reader *BitInputStream) byte
}

type Leaf struct {
	freq uint
	symbol byte
}

type Node struct {
	freq uint
	left Tree
	right Tree
}

type HuffTree struct {
	tree Tree
	codes []string
}

func NewLeaf(f uint, sym byte) *Leaf {
	p := new(Leaf)
	p.freq = f
	p.symbol = sym
	return p
}

func NewNode(f uint, l Tree, r Tree) *Node {
	p := new(Node)
	p.freq = l.Freq() + r.Freq()
	p.left = l
	p.right = r
	return p
}

func (l Leaf) Freq() uint {
	return l.freq
}

func (n Node) Freq() uint {
	return n.freq
}

func (l Leaf) WriteTo(writer *BitOutputStream) {
	writer.WriteBit(1)
	writer.WriteByte(uint16(l.symbol))
}

func (n Node) WriteTo(writer *BitOutputStream) {
	writer.WriteBit(0)
	n.left.WriteTo(writer)
	n.right.WriteTo(writer)
}

func (l Leaf) DumpCodes(codes []string, prefix string) {
	codes[l.symbol] = prefix
}

func (n Node) DumpCodes(codes []string, prefix string) {
	n.left.DumpCodes(codes, prefix+"0")
	n.right.DumpCodes(codes, prefix+"1")
}

func (l Leaf) ReadChar(reader *BitInputStream) byte {
	return l.symbol
}

func (n Node) ReadChar(reader *BitInputStream) byte {
	if reader.ReadBool() {
		return n.right.ReadChar(reader)
	} else {
		return n.left.ReadChar(reader)
	}
}

func BuildHuffTree(reader *BitInputStream) *HuffTree {
	p := new(HuffTree)

	freqs := calcFrecuencies(reader)
	heap := NewHeap(MAX_SYMBOLS)
	for s, f := range(freqs) {
		if f > 0 {
			heap.Insert(NewLeaf(uint(f), byte(s)))
		}
	}
	
	for heap.Size() > 1 {
		l := heap.Extract()
		r := heap.Extract()
		heap.Insert(NewNode(l.Freq()+r.Freq(), l, r))
	}
	tree := heap.Extract()
	codes := buildCodes(tree)
	p.tree = tree
	p.codes = codes
	return p
}

func buildCodes(tree Tree) []string {
	prefix := ""
	codes := make([]string, MAX_SYMBOLS)
	tree.DumpCodes(codes, prefix)
	return codes
}

func (h *HuffTree) Compress(reader *BitInputStream, writer *BitOutputStream) {
	h.tree.WriteTo(writer)
	h.writeSymbols(reader, writer)
	writer.Close()
}

func (h *HuffTree) Decompress(reader *BitInputStream, writer *BitOutputStream) {
	len := reader.ReadInt()
	fmt.Printf("decompress len = %d\n", len)
	for i := uint32(0); i < len; i++ {
		writer.WriteByte(uint16(h.tree.ReadChar(reader)))
	}
	writer.Close()
}

func (h HuffTree) writeSymbols(reader *BitInputStream, writer *BitOutputStream) {
	bytes := reader.GetBytes()
	lbytes := uint32(len(bytes))
	writer.WriteInt(lbytes)
	for _, s := range(bytes) {
		code := h.codes[s]
		for _, c := range(code) {
			writer.WriteBit(uint16(int(c) - '0'))
		}
	}
}


func ReadHuffTree(reader *BitInputStream) *HuffTree {
	p := new(HuffTree)
	p.tree = readTree(reader)
	return p	
}

func readTree(reader *BitInputStream) Tree {
	flag := reader.ReadBool()
	if flag {
		return *NewLeaf(0, reader.ReadChar())
	} else {
		l := readTree(reader)
		r := readTree(reader)
		return *NewNode(0, l, r)
	}
}

func calcFrecuencies(reader *BitInputStream) []uint {
	freqs := make([]uint, MAX_SYMBOLS)
	bytes := reader.GetBytes()
	for _, c := range(bytes) {
		freqs[c] += 1
	}
	return freqs
}

File: desafio4/swift/Sources/hufftree.swift
import Foundation

let MAX_SYMBOLS = 256

protocol Tree {
	
	func freq() -> UInt

	func writeTo(writer: BitOutputStream)

	func dumpCodes(codes: inout [String], prefix: String)

	func readChar(reader: BitInputStream) -> UInt8
}


class Leaf : Tree {
	
	var frecuency : UInt
	var symbol : UInt8

	init(frq:UInt, sym:UInt8) {
		frecuency = frq
		symbol = sym
	}

	func freq() -> UInt {
		return frecuency
	}

	func writeTo(writer: BitOutputStream) {
		writer.writeBit(bit: 1)
		writer.writeByte(byte: UInt16(symbol))
	}

	func dumpCodes(codes: inout [String], prefix: String) {
		codes[Int(symbol)] = prefix
	}

	func readChar(reader: BitInputStream) -> UInt8 {
		return symbol
	}
}


class Node : Tree {
	var left: Tree
	var right: Tree

	init( left: Tree, right: Tree) {
		self.left = left
		self.right = right
	}	

	func freq() -> UInt {
		return left.freq() + right.freq()
	}

	func writeTo(writer: BitOutputStream) {
		writer.writeBit(bit: 0)
		left.writeTo(writer: writer)
		right.writeTo(writer: writer)
	}

	func dumpCodes(codes: inout [String], prefix: String) {
		left.dumpCodes(codes: &codes, prefix: prefix+"0")
		right.dumpCodes(codes: &codes, prefix: prefix+"1")
	}

	func readChar(reader: BitInputStream) -> UInt8 {
		if reader.readBool() {
			return right.readChar(reader: reader)
		} else {
			return left.readChar(reader: reader)
		}
	}
}

class HuffTree {


	var tree: Tree? = nil 
	var codes: [String] = [String]()

	init(buildFrom: BitInputStream) {
		let freqs = calcFrecuencies(source: buildFrom)
		let heap = Heap(size: MAX_SYMBOLS)

		for (s, f) in freqs.enumerated() { 
			if f > 0 {
				heap.insert(elem: Leaf(frq: UInt(f), sym: UInt8(s)))
			}
		}

		while heap.size() > 1 {
			let l = heap.extract()
			let r = heap.extract()
			heap.insert(elem: Node(left: l!, right: r!))
		}
		tree = heap.extract()
		codes = buildCodes(tree: tree!)

	}

	init(readFrom: BitInputStream) {
		tree = readTree(reader: readFrom)
	}

	func compress(readFrom: BitInputStream, writeTo: BitOutputStream) {
		tree!.writeTo(writer: writeTo)
		writeSymbols(reader: readFrom, writer: writeTo)
		writeTo.close()
	}

	func decompress(readFrom: BitInputStream, writeTo: BitOutputStream) {
		let len = readFrom.readInt()
		for _ in 0..<len {
			writeTo.writeByte(byte: UInt16(tree!.readChar(reader: readFrom)))
		}
		writeTo.close()
	}

	func readTree(reader: BitInputStream) -> Tree {
		let flag = reader.readBool()
		if flag {
			return Leaf(frq: 0, sym: reader.readChar())
		} else {
			let l = readTree(reader: reader)
			let r = readTree(reader: reader)
			return Node(left: l, right: r)
		}
	}

	func calcFrecuencies(source: BitInputStream) -> [Int]{
		var freqs = [Int](repeating: 0, count: MAX_SYMBOLS)
		let bytes = source.getBytes()
		for i in 0 ..< bytes.count {
			let c = bytes[i]
			freqs[Int(c)] += 1
		}
		return freqs
	}

	func buildCodes(tree: Tree) -> [String] {
		var codes = [String](repeating: "", count: MAX_SYMBOLS)
		tree.dumpCodes(codes: &codes, prefix: "")
		return codes
	}

	func writeSymbols(reader: BitInputStream, writer: BitOutputStream) {
		let bytes = reader.getBytes()
		let lbytes = bytes.count
		writer.writeInt(i: UInt32(lbytes))
		for (_, s) in bytes.enumerated() {
			let code = codes[Int(s)]
			for c in code {
				writer.writeBitc(bit: c)
			}
		}
	}
}